src/utils/callGraph.ts

```diff
@@ -0,0 +1,163 @@
+    const parent = this.findNode(parentId, this.root);
+    /* Step 2: Check if a child node to parent with same name exists; 
+    if yes, then append to the events property of that node
+    else, create new Node, add to the events property and append to children of parent
+    */
+    if (parent) {
+      parent.children.push(new Node(parseInt(nodeId), []));
+      return true;
+    }
+    return false;
+  }
+
+  /* Add the sample events to the nodeId corresponding to their stackFrame */
+  addEvent(event: HermesSample): boolean {
+    const stackFrameNode = this.findNode(event.sf.toString(), this.root);
+    if (stackFrameNode) {
+      stackFrameNode.events.push(event);
+      return true;
+    }
+    return false;
+  }
+
+  /* Can give the state of the graph below a particular node at any time */
+  printAllNodeIds(node: Node): void {
+    console.log(this.findNode(node.id.toString(), this.root));
+    for (let i = 0; i < node.children.length; i++) {
+      this.printAllNodeIds(node.children[i]);
+    }
+  }
+
+  /* Get list of events by traversing the graph */
+  listOfEvents(node: Node): HermesSample[] {
+    const events: HermesSample[] = [];
+    events.push(...node.events);
+    for (let i = 0; i < node.children.length; i++) {
+      events.push(...this.listOfEvents(node.children[i]));
+    }
+    return events;
+  }
+
+  /* Sort Node Events by Timestamp */
+  sortNodeEvents(node: Node): Node {
+    node.events = node.events.sort((a, b) =>
+      a.ts < b.ts ? -1 : a.ts > b.ts ? 1 : 0
+    );
+    return node;
+  }
+
+  constructChromeSample = (
+    sample: HermesSample,
+    ph: EventsPhase
+  ): SharedEventProperties => {
+    return {
+      pid: sample.pid,
+      tid: Number(sample.tid),
+      ts: Number(sample.ts),
+      name: sample.name,
+      cat: sample.stackFrameData?.category || 'JavaScript',
+      args: { name: sample.name },
+      ph: ph,
+    };
+  };
+
+  /*Remove Events with incorrect number; 
+  i.e. if the same event name has an odd number of events associated to it,
+   make the last instance an instant event rather than a Duration Event */
+  fixChromeEvents(
+    chromeEvents: SharedEventProperties[],
+    eventDetails: { [key in string]: { ph: EventsPhase; count: number } }
+  ): SharedEventProperties[] {
+    const incorrectEvents: string[] = [];
+    for (let eventName in eventDetails) {
+      if (eventDetails[eventName].count % 2) {
+        incorrectEvents.push(eventName);
+      }
+    }
+    const correctEvents: SharedEventProperties[] = [];
+    if (incorrectEvents.length) {
+      for (let i = chromeEvents.length - 1; i >= 0; i--) {
+        if (incorrectEvents.includes(chromeEvents[i].name!)) {
+          _.remove(incorrectEvents, event => event === chromeEvents[i].name);
+        }
+        correctEvents.push(chromeEvents[i]);
+      }
+    }
+    return correctEvents;
+  }
+
+  /* Sets phase for all events in one node */
+  setNodePhase(node: Node): SharedEventProperties[] {
```

Comment:
This method assigns a phase to each sample event. We assign the phase using a simple peek method, wherein if we haven't seen the function name before we assign it a Duration Begin and if we have we assign it a Duration End.

There is a lot of scope of improvement here as we only use Duration Events here and we ideally would want to use the whole range of events listed in the trace event format document.

Reply:
Edit: After meeting with Ram, what @jessieAnhNguyen and I understand is that, it's better to shift to a Complete Event format, because contrary to what we believed earlier, the sample simply contain like snapshots of events.
So it makes sense to move to Complete events with duration fields, for the following reason/s
1) A duration property is easier to manage.
2) Instead of the peek approach, we can not just check for the earlier and the latest instances of the events in each node and convert those to events. There is no need to worry about any other occurences, as the samples simply represent snapshots of events in the stack and not actual function calls.